refactor(writer): avoid some intermediate allocations

fogti opened the pull request:
### Summary of the PR

This PR reduces the amount of intermediate allocations done in `property_string` and `property_string_list`.
This doesn't change the public interface at all, but should also make it easier to avoid intermediate allocations in the future. (see also PR #76 as something it might be utilized in conjunction with).

Background: `vm-fdt` is used by https://github.com/hermit-os/loader/blob/0c1c1a819fa77016672a620ebaa44e79b8c95599/src/fdt.rs, which is an environment where we want to avoid allocations, particularly of small, intermediate buffers as much as possible, particularly when such buffers aren't a necessity and can be avoided relatively easily.

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [X] All commits in this PR have Signed-Off-By trailers (with
  `git commit -s`), and the commit message has max 60 characters for the
  summary and max 75 characters for each description line.
- [X] All added/changed functionality has a corresponding unit/integration
  test.
- [X] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [X] Any newly added `unsafe` code is properly documented.

fogti commented on src/writer.rs:
```diff
@@ -447,16 +490,22 @@ impl FdtWriter {
-        let cstr_value = CString::new(val).map_err(|_| Error::InvalidString)?;
-        self.property(name, cstr_value.to_bytes_with_nul())
+        self.property_and_nul(
+            &CString::new(name).map_err(|_| Error::InvalidString)?,
```
This is done mostly with PR #75 in mind and can be moved there if deemed suitable.

danielverkamp replied:
We end up allocating a `CString` here and then allocating another one inside `intern_string()` in the case this is a new string; maybe it would be possible to change the API so that `intern_string()` expects the string without a NUL terminator (in the same way as `property_and_nul()`) so it would only get allocated once?

danielverkamp commented on src/writer.rs:
```diff
@@ -216,21 +218,28 @@ fn node_name_valid_first_char(c: char) -> bool {
 // Check if `name` is a valid property name.
 // https://devicetree-specification.readthedocs.io/en/stable/devicetree-basics.html#property-names
 fn property_name_valid(name: &str) -> bool {
```
Since `property_name_valid()` is an internal function (not part of the public API), it should be safe to remove it (and `property_name_valid_char()`) and replace its usage with the new `&CStr` version, rather than duplicating the list of valid characters.

The only `property_name_valid()` is called is in `property()`, which already creates a `CString`, so it could be updated to something like:

```
let name_cstr = ...;

if !property_cstr_name_valid(&name_cstr) {
    return Err(Error::InvalidPropertyName);
}
```

fogti commented:
Could we move this forward, what's missing?

rbradford commented:
> Could we move this forward, what's missing?

Looks like the CI failed due to coverage change. If you update your coverage % and force push that should clear and can move forward

fogti committed:
refactor(writer): avoid some intermediate allocations

This doesn't change the public interface at all, but should also make it easier
to avoid intermediate allocations in the future.

Signed-off-by: Έλλεν Εμίλια Άννα Zscheile <fogti+devel@ytrizja.de>

fogti commented:
Some other CI pipelines are failing now, but they seem unrelated to the PR.
